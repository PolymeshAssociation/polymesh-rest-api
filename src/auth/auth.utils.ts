import { AuthGuard, IAuthGuard } from '@nestjs/passport';

import { AuthStrategy, authStrategyValues } from '~/auth/strategies/strategies.consts';

/**
 *  Creates an AuthGuard using the configured strategies
 */
export const createAuthGuard = (rawStrategy: string): IAuthGuard => {
  const strategies = parseAuthStrategyConfig(rawStrategy);
  return new (class extends AuthGuard(strategies) {})();
};

/**
 * transforms a raw auth strategy config into valid strategy values
 *
 * @throws if given invalid values
 */
const parseAuthStrategyConfig = (rawStrategyConfig: string): AuthStrategy[] => {
  const givenStrategies = rawStrategyConfig.split(',').map(strategy => strategy.trim());

  const filteredStrategies = givenStrategies.filter(isStrategyKey);

  if (filteredStrategies.length !== givenStrategies.length) {
    throw new Error(
      `Auth config error! "${givenStrategies}" contains an unrecognized option. Valid values are: "${authStrategyValues}"`
    );
  }

  return filteredStrategies.sort(cmpAuthStrategyOrder);
};

export const parseApiKeysConfig = (rawApiKeyConfig: string): string[] => {
  if (rawApiKeyConfig.trim() === '') {
    return [];
  }

  return rawApiKeyConfig.split(',').map(rawKey => rawKey.trim());
};

const isStrategyKey = (key: string): key is AuthStrategy => {
  return authStrategyValues.includes(key as AuthStrategy);
};

/**
 * A helper to be passed to `.sort`. The order is in which they will be evaluated.
 *
 * For example the the "open" strategy should always be last to allow for testing of other strategies
 */
const cmpAuthStrategyOrder = (a: AuthStrategy, b: AuthStrategy): number =>
  authStrategyValues.indexOf(a) - authStrategyValues.indexOf(b);
