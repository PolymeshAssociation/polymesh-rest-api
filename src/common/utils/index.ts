import { ModuleName, TxTags } from '@polymeshassociation/polymesh-sdk/types';
import { randomBytes } from 'crypto';
import { flatten } from 'lodash';
import { promisify } from 'util';

import { TransactionBaseDto } from '~/common/dto/transaction-base-dto';
import { NotificationPayloadModel } from '~/common/models/notification-payload-model';
import { TransactionQueueModel } from '~/common/models/transaction-queue.model';
import { EventType } from '~/events/types';
import { NotificationPayload } from '~/notifications/types';
import { TransactionResult } from '~/transactions/transactions.util';

/* istanbul ignore next */
export function getTxTags(): string[] {
  return flatten(Object.values(TxTags).map(txTag => Object.values(txTag)));
}

/* istanbul ignore next */
export function getTxTagsWithModuleNames(): string[] {
  const txTags = getTxTags();
  const moduleNames = Object.values(ModuleName);
  return [...moduleNames, ...txTags];
}

export type TransactionResponseModel = NotificationPayloadModel | TransactionQueueModel;

/**
 * A helper type that lets a service return a QueueResult or a Subscription Receipt
 */
export type ServiceReturn<T> = Promise<
  NotificationPayload<EventType.TransactionUpdate> | TransactionResult<T>
>;

/**
 * A helper type that lets a controller return a Model or a Subscription Receipt if webhookUrl is being used
 */
export type TransactionResolver<T> = (
  res: TransactionResult<T>
) => Promise<TransactionQueueModel> | TransactionQueueModel;

/**
 * A helper function that transforms a service result for a controller. A controller can pass a resolver for a detailed return model, otherwise the transaction details will be used as a default
 */
export const handleServiceResult = <T>(
  result: NotificationPayloadModel | TransactionResult<T>,
  resolver: TransactionResolver<T> = basicModelResolver
): NotificationPayloadModel | Promise<TransactionQueueModel> | TransactionQueueModel => {
  if ('transactions' in result) {
    return resolver(result);
  }

  return new NotificationPayloadModel(result);
};

/**
 * A helper function for controllers when they should return a basic TransactionQueueModel
 */
const basicModelResolver: TransactionResolver<unknown> = ({ transactions, details }) => {
  return new TransactionQueueModel({ transactions, details });
};

/**
 * Generate base64 encoded, cryptographically random bytes
 *
 * @note random byte length given, not the encoded string length
 */
export const generateBase64Secret = async (byteLength: number): Promise<string> => {
  const buf = await promisify(randomBytes)(byteLength);

  return buf.toString('base64');
};

/**
 * Helper class to ensure a code path is unreachable. For example this can be used for ensuring switch statements are exhaustive
 */
export class UnreachableCaseError extends Error {
  /** This should never be called */
  constructor(val: never) {
    super(`Unreachable case: ${JSON.stringify(val)}`);
  }
}

export const extractTxBase = <T extends TransactionBaseDto>(
  params: T
): {
  base: TransactionBaseDto;
  args: Omit<T, keyof TransactionBaseDto>;
} => {
  const { signer, webhookUrl, dryRun, ...args } = params;
  return {
    base: { signer, webhookUrl, dryRun },
    args,
  };
};
