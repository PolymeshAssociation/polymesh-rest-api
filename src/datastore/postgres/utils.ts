/* istanbul ignore file */

export interface PostgresConfig {
  type: 'postgres';
  host: string;
  username: string;
  password: string;
  database: string;
  port: number;
}

export const readPostgresConfigFromEnv = (): PostgresConfig | undefined => {
  const {
    REST_POSTGRES_HOST: host,
    REST_POSTGRES_PORT: port,
    REST_POSTGRES_USER: username,
    REST_POSTGRES_PASSWORD: password,
    REST_POSTGRES_DATABASE: database,
  } = process.env;

  if (!host || !port || !username || !password || !database) {
    return undefined;
  }

  return { type: 'postgres', host, username, port: Number(port), password, database };
};
