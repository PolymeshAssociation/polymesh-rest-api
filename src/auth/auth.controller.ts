import { Body, Controller, Post } from '@nestjs/common';
import { ApiNoContentResponse, ApiOkResponse, ApiOperation, ApiTags } from '@nestjs/swagger';

import { AuthService } from '~/auth/auth.service';
import { CreateApiKeyDto } from '~/auth/dto/create-api-key.dto';
import { DeleteApiKeyDto } from '~/auth/dto/delete-api-key.dto';
import { ApiKeyModel } from '~/auth/models/api-key.model';

@ApiTags('auth')
@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @ApiOperation({
    summary: 'Create API Key',
    description: 'This endpoint will create an API Key',
  })
  @ApiOkResponse({
    description: 'Details of the API key created',
    type: ApiKeyModel,
  })
  @Post('api-key/create')
  public async createApiKey(@Body() params: CreateApiKeyDto): Promise<ApiKeyModel> {
    return this.authService.createApiKey(params);
  }

  @ApiOperation({
    summary: 'Delete an API Key',
    description: 'This endpoint invalidates the given API key',
  })
  @ApiNoContentResponse({
    description: 'The API key is no longer valid',
  })
  @Post('/api-key/delete')
  public async deleteApiKey(@Body() params: DeleteApiKeyDto): Promise<void> {
    await this.authService.deleteApiKey(params);
  }
}
