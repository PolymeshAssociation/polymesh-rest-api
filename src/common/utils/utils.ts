import { BadRequestException, InternalServerErrorException } from '@nestjs/common';
import {
  ErrorCode,
  isPolymeshError,
  ProcedureMethod,
  ProcedureOpts,
} from '@polymathnetwork/polymesh-sdk/types';

import { QueueResult } from '~/common/types';

export async function processQueue<MethodArgs, ReturnType>(
  method: ProcedureMethod<MethodArgs, unknown, ReturnType>,
  args: MethodArgs,
  opts: ProcedureOpts
): Promise<QueueResult<ReturnType>> {
  // TODO @monitz87: Improve error handling
  try {
    const queue = await method(args, opts);
    const result = await queue.run();

    return {
      result,
      transactions: queue.transactions.map(({ blockHash, txHash, tag }) => ({
        /* eslint-disable @typescript-eslint/no-non-null-assertion */
        blockHash: blockHash!,
        transactionHash: txHash!,
        transactionTag: tag,
        /* eslint-enable @typescript-eslint/no-non-null-assertion */
      })),
    };
  } catch (err) /* istanbul ignore next: not worth the trouble */ {
    if (isPolymeshError(err)) {
      const { message, code } = err;
      switch (code) {
        case ErrorCode.ValidationError:
          throw new BadRequestException(message);
        default:
          throw new InternalServerErrorException(message);
      }
    }
    throw new InternalServerErrorException(err.message);
  }
}
