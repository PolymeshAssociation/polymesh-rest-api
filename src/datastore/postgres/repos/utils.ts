import { TypeORMError } from 'typeorm';

import { AppConflictError } from '~/common/errors';

const isTypeOrmError = (err: Error): err is TypeORMError => {
  return err instanceof TypeORMError;
};

export const convertTypeOrmErrorToAppError =
  (id: string, resourceType: string) =>
  (err: Error): void => {
    if (isTypeOrmError(err)) {
      const { message } = err;
      if (message.includes('duplicate key value violates unique constraint')) {
        throw new AppConflictError(id, resourceType);
      }
    }

    throw err;
  };
