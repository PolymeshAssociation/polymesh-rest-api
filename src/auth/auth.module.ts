/* istanbul ignore file */

import { Module } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { APP_GUARD } from '@nestjs/core';
import { IAuthGuard, PassportModule } from '@nestjs/passport';

import { AuthController } from '~/auth/auth.controller';
import { AuthService } from '~/auth/auth.service';
import { createAuthGuard } from '~/auth/auth.utils';
import { ApiKeyStrategy } from '~/auth/strategies/api-key.strategy';
import { OpenStrategy } from '~/auth/strategies/open.strategy';
import { DatastoreModule } from '~/datastore/datastore.module';
import { UsersModule } from '~/users/users.module';

/**
 * responsible for the REST API's authentication strategies
 *
 * @note authorization has not yet been implemented - all users have full access
 */
@Module({
  imports: [
    ConfigModule,
    DatastoreModule.registerAsync(),
    UsersModule,
    PassportModule.register({
      session: false,
    }),
  ],
  providers: [
    AuthService,
    ApiKeyStrategy,
    OpenStrategy,
    {
      provide: APP_GUARD, // registers a global guard
      useFactory: (config: ConfigService): IAuthGuard => {
        const configuredStrategies = config.getOrThrow('AUTH_STRATEGY');
        return createAuthGuard(configuredStrategies);
      },
      inject: [ConfigService],
    },
  ],
  controllers: [AuthController],
  exports: [AuthService, PassportModule],
})
export class AuthModule {}
