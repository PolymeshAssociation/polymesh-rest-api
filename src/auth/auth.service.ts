import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { parseApiKeysConfig } from '~/auth/auth.utils';
import { CreateApiKeyDto } from '~/auth/dto/create-api-key.dto';
import { DeleteApiKeyDto } from '~/auth/dto/delete-api-key.dto';
import { ApiKeyModel } from '~/auth/models/api-key.model';
import { generateBase64Secret } from '~/common/utils';
import { User, UserId } from '~/users/user.consts';

@Injectable()
export class AuthService {
  private userStore: Record<UserId, User> = {};

  constructor(config: ConfigService) {
    const givenApiKeys = config.getOrThrow<string>('API_KEYS');
    const apiKeys = parseApiKeysConfig(givenApiKeys);

    const configuredUserId = 'configured-user';
    this.userStore[configuredUserId] = { id: configuredUserId, apiKeys };
  }

  public async createApiKey({ userId }: CreateApiKeyDto): Promise<ApiKeyModel> {
    const secret = await generateBase64Secret(32);

    if (!this.userStore[userId]) {
      this.userStore[userId] = { id: userId, apiKeys: [] };
    }
    this.userStore[userId].apiKeys.push(secret);

    return { userId, secret };
  }

  /**
   * Returns the associated user by an API key
   */
  public validateApiKey(apiKey: string): User | undefined {
    return this.getUserByApiKey(apiKey);
  }

  /**
   * @param apiKey - key to delete
   */
  public async deleteApiKey({ apiKey }: DeleteApiKeyDto): Promise<void> {
    const user = this.getUserByApiKey(apiKey);
    if (user) {
      user.apiKeys = user.apiKeys.filter(key => key !== apiKey);
    }
  }

  /**
   * @note, will likely need to be a abstracted with a "repository" type interface when a data store is added
   */
  private getUserByApiKey(apiKey: string): User | undefined {
    return Object.values(this.userStore).find(({ apiKeys }) => apiKeys.includes(apiKey));
  }
}
