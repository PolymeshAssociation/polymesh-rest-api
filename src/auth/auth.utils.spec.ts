import { createAuthGuard, parseApiKeysConfig } from '~/auth/auth.utils';

describe('createAuthGuard', () => {
  it('should handle a single option', async () => {
    const guard = createAuthGuard('apiKey');
    expect(guard).toBeDefined();
  });

  it('should handle multiple valid options', () => {
    const guard = createAuthGuard('apiKey,open');
    expect(guard).toBeDefined();
  });

  it('should throw if an invalid option is given', () => {
    return expect(() =>
      createAuthGuard('open,apiKey,NOT_A_STRATEGY')
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('parseApiKeysConfig', () => {
  it('should split and trim on commas', () => {
    const result = parseApiKeysConfig('abc,def, ghi ');
    expect(result).toEqual(['abc', 'def', 'ghi']);
  });
});
