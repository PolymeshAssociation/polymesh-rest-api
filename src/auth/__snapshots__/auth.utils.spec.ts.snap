// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAuthGuard should throw if an invalid option is given 1`] = `"Auth config error! "open,apiKey,NOT_A_STRATEGY" contains an unrecognized option. Valid values are: "apiKey,open""`;
